datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs

```diff
@@ -281,7 +280,13 @@ impl<'a> TreeNodeRewriter for ConstEvaluator<'a> {
 
     fn mutate(&mut self, expr: Expr) -> Result<Expr> {
         match self.can_evaluate.pop() {
-            Some(true) => Ok(Expr::Literal(self.evaluate_to_scalar(expr)?)),
+            Some(true) => {
+                let lit = self.evaluate_to_scalar(expr.clone());
```

Comment:
It seems a real shame that this has to clone the expression even for the common case where there is no error

 maybe we could use something like the following:
```rust
enum SimplifyResult {
   // expr was simplifed and contains the new expression
  Simplified(Expr),
  // Evalaution encountered an error, contains the original expression
  SimplifyRuntimeError(DataFusionError, Expr), 
}
```